### PR TITLE
new layer to store charge states of SiDBs, tests included

### PR DESCRIPTION
**Pull request by Drewniok**

### CODE REVIEW

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ */
+enum class sidb_charge_states
+{
+    NONE,  // assigned when layout cell is empty
+    POSITIVE,
+    NEUTRAL,
+    NEGATIVE
+};
+
+struct sidb_charge
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:28]: We don't need this struct. In `charge_distribution_surface`, we can map directly from `coordinate` to `sidb_charge_state`

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+#include <unordered_map>
+
+namespace fiction
+{
+
+/**
+ * Possible SiDB charges.
+ */
+enum class sidb_charge_states
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:20]: I think I would go with singular `state`, but that's just personal preference

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    const sidb_charge_states charge_state;
+};
+
+template <typename Lyt>
+class charge_distribution_surface : public Lyt
+{
+  public:
+    struct charge_distribution_storage
+    {
+        std::unordered_map<coordinate<Lyt>, sidb_charge> charge_coordinates{};
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:46]: see above

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * Standard constructor.
+     */
+    constexpr explicit sidb_charge(const sidb_charge_states defect_type = sidb_charge_states::NONE) noexcept :
+            charge_state{defect_type}
+    {}
+
+    const sidb_charge_states charge_state;
+};
+
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:39]: Please add nested overloads as in `sidb_surface` to optimize the use of `charge_distribution_surface<charge_distribution_surface>`. You can create according traits in `traits.hpp`

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+template <typename Lyt>
+class charge_distribution_surface : public Lyt
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:41]: Please add documentation to the class

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    };
+
+    using storage = std::shared_ptr<charge_distribution_storage>;
+
+    /**
+     * Standard constructor for empty layouts.
+     */
+    explicit charge_distribution_surface() : Lyt(), strg{std::make_shared<charge_distribution_storage>()}
+    {
+
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:56]: Superfluous empty line

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+     */
+    void assign_charge_state(const coordinate<Lyt>& c, const sidb_charge& cs) noexcept
+    {
+        if (!Lyt::is_empty_cell(c))
+        {
+            if (const auto it = strg->charge_coordinates.find(c); it != strg->charge_coordinates.cend())
+            {
+                strg->charge_coordinates.erase(c);
+                strg->charge_coordinates.insert({c, cs});
+            }
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:84]: You don't need this bit. The `insert` function will take care of that. You could check if `cs` is `NONE` and then remove the stored charge accordinly

File: test/technology/charge_distribution_surface.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
+//
+// Created by Jan Drewniok on 05.12.22.
+//
+//
+// Created by marcel on 07.03.22.
```
Comment (marcelwa) [inline test/technology/charge_distribution_surface.cpp:5]: I don't think, I was involved here :smile:

File: test/technology/charge_distribution_surface.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    (sidb_surface<
+        cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, even_column_hex>>>>))
+
+{
+    TestType                    lyt{{11, 9}};
+    charge_distribution_surface charge_layout{lyt};
+
+    SECTION("assign and read out charge states")
+    {
+
```
Comment (marcelwa) [inline test/technology/charge_distribution_surface.cpp:43]: missing comment like below?

File: include/fiction/traits.hpp [hunk 10]
```diff
@@ -697,6 +697,50 @@ template <class Lyt>
...
+struct has_get_charge_state : std::false_type
+{};
+
+template <class Lyt>
+struct has_get_charge_state<Lyt, std::void_t<decltype(std::declval<Lyt>().get_charge_state(coordinate<Lyt>()))>>
+        : std::true_type
+{};
+
+template <class Lyt>
+inline constexpr bool has_get_charge_state_v = has_get_sidb_defect<Lyt>::value;
```
Comment (marcelwa) [inline include/fiction/traits.hpp:726]: This line was probably supposed to be `inline constexpr bool has_get_charge_state_v = has_get_charge_state<Lyt>::value;`

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 11]
```diff
@@ -69,7 +84,7 @@ class charge_distribution_surface : public Lyt
     {
         if (!Lyt::is_empty_cell(c) && cs != sidb_charge_state::NONE)
         {
-            strg->charge_coordinates.insert_or_assign(c,cs);
+            strg->charge_coordinates.insert_or_assign(c, cs);
         }
         else if (!Lyt::is_empty_cell(c) && cs == sidb_charge_state::NONE)
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:89]: It is not really necessary to check for `!Lyt::is_empty_cell(c)` twice. You could do one big outer `if` and then do another `if-else` inside. Would prevent evaluating the same condition two times.

File: include/fiction/traits.hpp [hunk 12]
```diff
@@ -697,6 +697,50 @@ template <class Lyt>
...
+template <class Lyt>
+struct has_foreach_charge_state<Lyt, std::void_t<decltype(std::declval<Lyt>().foreach_charge_state(
+                                         std::declval<void(std::pair<coordinate<Lyt>, sidb_defect>, uint32_t)>()))>>
+        : std::true_type
+{};
+
+template <class Lyt>
+inline constexpr bool has_foreach_charge_state_v = has_foreach_charge_state<Lyt>::value;
+#pragma endregion
+
```
Comment (marcelwa) [inline include/fiction/traits.hpp:743]: Could you please add tests for the traits as well? See `sidb_surface.cpp` for examples

**Review by marcelwa — CHANGES_REQUESTED** [review]

File: include/fiction/technology/charge_distribution_surface.hpp (hunk 11 above)
Comment (Drewniok) [reply include/fiction/technology/charge_distribution_surface.hpp:89]: corrected

File: include/fiction/traits.hpp (hunk 10 above)
Comment (Drewniok) [reply include/fiction/traits.hpp:726]: You are right.

**Comment by marcelwa** [comment]
I guess there is a missing `#include <cstdint>` in `sidb_surface.hpp`, causing the `clang-format` errors. That one is on me. If you could add it, that would be nice :)

File: include/fiction/technology/charge_distribution_surface.hpp (hunk 11 above)
Comment (marcelwa) [reply include/fiction/technology/charge_distribution_surface.hpp:79]: Not yet :sweat_smile:

File: test/technology/charge_distribution_surface.cpp [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<cartesian_layout<offset::ucoord_t>>>>),
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, odd_row_hex>>>>),
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, even_row_hex>>>>),
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, odd_column_hex>>>>),
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, even_column_hex>>>>))
+{
+    REQUIRE(is_cell_level_layout_v<TestType>);
+    CHECK(has_assign_sidb_defect_v<TestType>);
+    CHECK(has_get_sidb_defect_v<TestType>);
+    CHECK(has_foreach_sidb_defect_v<TestType>);
```
Comment (marcelwa) [inline test/technology/charge_distribution_surface.cpp:27]: It doesn't make much sense to test the defects here. Rather check that the charge state interface is not present.

File: test/technology/charge_distribution_surface.cpp [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+using namespace fiction;
+
+TEMPLATE_TEST_CASE(
+    "charge distribution surface traits and construction", "[sidb-surface]",
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<cartesian_layout<offset::ucoord_t>>>>),
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, odd_row_hex>>>>),
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, even_row_hex>>>>),
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, odd_column_hex>>>>),
+    (sidb_surface<cell_level_layout<sidb_technology, clocked_layout<hexagonal_layout<offset::ucoord_t, even_column_hex>>>>))
```
Comment (marcelwa) [inline test/technology/charge_distribution_surface.cpp:22]: Can we also add types that do not implement the `sidb_surface`?

**Review by marcelwa — CHANGES_REQUESTED** [review]

**Review by marcelwa — APPROVED** [review]

**Pull request merged**